File: src/runtime/modules/shu/lib/tls/tls.c
/**
 * TLS 服务端实现：OpenSSL SSL_CTX + SSL_accept/read/write。
 * 编译需链接 -lssl -lcrypto。
 * BIO 模式：用于 Windows IOCP，握手与连接读写由调用方用 overlapped recv/send 驱动。
 */
#include "tls.h"
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <openssl/bio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

struct tls_ctx {
    SSL_CTX *ctx;
};

struct tls_conn {
    SSL *ssl;
    int fd;
    int last_ret; /* 上次 SSL_read/SSL_write 返回值，供 want_read/want_write */
};

/** 非阻塞握手中：持有 SSL 未完成握手，由 tls_accept_start 创建、tls_accept_step 推进或 tls_pending_free 释放 */
struct tls_pending {
    SSL_CTX *ctx;
    SSL *ssl;
    int fd;
    int last_ret; /* SSL_accept 上次返回值，供 want_read/want_write 用 */
};

/** 客户端 TLS 上下文：仅持有 SSL_CTX，用于 tls_connect */
struct tls_client_ctx {
    SSL_CTX *ctx;
};

tls_ctx_t *tls_ctx_create(const char *cert_path, const char *key_path) {
    const SSL_METHOD *method = TLS_server_method();
    if (!method) return NULL;
    SSL_CTX *ctx = SSL_CTX_new(method);
    if (!ctx) return NULL;
    if (SSL_CTX_set_min_proto_version(ctx, TLS1_2_VERSION) != 1) {
        SSL_CTX_free(ctx);
        return NULL;
    }
    if (SSL_CTX_use_certificate_file(ctx, cert_path, SSL_FILETYPE_PEM) != 1) {
        SSL_CTX_free(ctx);
        return NULL;
    }
    if (SSL_CTX_use_PrivateKey_file(ctx, key_path, SSL_FILETYPE_PEM) != 1) {
        SSL_CTX_free(ctx);
        return NULL;
    }
    if (SSL_CTX_check_private_key(ctx) != 1) {
        SSL_CTX_free(ctx);
        return NULL;
    }
    /* ALPN: 服务端通告 h2，握手后可用 tls_get_alpn_selected 判断是否走 HTTP/2 */
    static const unsigned char alpn_protos[] = { 2, 'h', '2' };
    if (SSL_CTX_set_alpn_protos(ctx, alpn_protos, sizeof(alpn_protos)) != 0) {
        SSL_CTX_free(ctx);
        return NULL;
    }
    tls_ctx_t *t = (tls_ctx_t *)malloc(sizeof(tls_ctx_t));
    if (!t) {
        SSL_CTX_free(ctx);
        return NULL;
    }
    t->ctx = ctx;
    return t;
}

void tls_ctx_free(tls_ctx_t *t) {
    if (!t) return;
    if (t->ctx) SSL_CTX_free(t->ctx);
    free(t);
}

tls_conn_t *tls_accept(tls_ctx_t *t, int fd) {
    if (!t || fd < 0) return NULL;
    SSL *ssl = SSL_new(t->ctx);
    if (!ssl) return NULL;
    if (SSL_set_fd(ssl, fd) != 1) {
        SSL_free(ssl);
        return NULL;
    }
    if (SSL_accept(ssl) <= 0) {
        SSL_free(ssl);
        return NULL;
    }
    tls_conn_t *conn = (tls_conn_t *)malloc(sizeof(tls_conn_t));
    if (!conn) {
        SSL_free(ssl);
        return NULL;
    }
    conn->ssl = ssl;
    conn->fd = fd;
    conn->last_ret = 0;
    return conn;
}

/* 非阻塞握手：调用方须已将 fd 设为 non-blocking */
tls_pending_t *tls_accept_start(tls_ctx_t *t, int fd) {
    if (!t || fd < 0) return NULL;
    SSL *ssl = SSL_new(t->ctx);
    if (!ssl) return NULL;
    if (SSL_set_fd(ssl, fd) != 1) {
        SSL_free(ssl);
        return NULL;
    }
    tls_pending_t *p = (tls_pending_t *)malloc(sizeof(tls_pending_t));
    if (!p) {
        SSL_free(ssl);
        return NULL;
    }
    p->ctx = t->ctx;
    p->ssl = ssl;
    p->fd = fd;
    p->last_ret = 0;
    return p;
}

/* 返回 1=成功(out_conn 已写入并释放 pending)，0=需 poll 后再调，-1=错误 */
int tls_accept_step(tls_pending_t *pending, tls_conn_t **out_conn) {
    if (!pending || !pending->ssl) return -1;
    int r = SSL_accept(pending->ssl);
    pending->last_ret = r;
    if (r == 1) {
        tls_conn_t *conn = (tls_conn_t *)malloc(sizeof(tls_conn_t));
        if (!conn) return -1;
        conn->ssl = pending->ssl;
        conn->fd = pending->fd;
        conn->last_ret = 0;
        pending->ssl = NULL;
        tls_pending_free(pending);
        if (out_conn) *out_conn = conn;
        return 1;
    }
    switch (SSL_get_error(pending->ssl, r)) {
        case SSL_ERROR_WANT_READ:
        case SSL_ERROR_WANT_WRITE:
            return 0;
        default:
            return -1;
    }
}

int tls_pending_want_read(tls_pending_t *pending) {
    if (!pending || !pending->ssl) return 0;
    return (SSL_get_error(pending->ssl, pending->last_ret) == SSL_ERROR_WANT_READ) ? 1 : 0;
}

int tls_pending_want_write(tls_pending_t *pending) {
    if (!pending || !pending->ssl) return 0;
    return (SSL_get_error(pending->ssl, pending->last_ret) == SSL_ERROR_WANT_WRITE) ? 1 : 0;
}

void tls_pending_free(tls_pending_t *pending) {
    if (!pending) return;
    if (pending->ssl) {
        SSL_free(pending->ssl);
        pending->ssl = NULL;
    }
    pending->fd = -1;
    free(pending);
}

int tls_read(tls_conn_t *conn, void *buf, int len) {
    if (!conn || !conn->ssl || !buf || len <= 0) return -1;
    int n = SSL_read(conn->ssl, buf, len);
    conn->last_ret = n;
    if (n > 0) return n;
    if (n == 0) return (SSL_get_error(conn->ssl, 0) == SSL_ERROR_ZERO_RETURN) ? 0 : -1;
    switch (SSL_get_error(conn->ssl, n)) {
        case SSL_ERROR_WANT_READ:  return -2;
        case SSL_ERROR_WANT_WRITE: return -3;
        case SSL_ERROR_ZERO_RETURN: return 0;
        default: return -1;
    }
}

int tls_write(tls_conn_t *conn, const void *buf, int len) {
    if (!conn || !conn->ssl || !buf || len <= 0) return -1;
    int n = SSL_write(conn->ssl, buf, (int)len);
    conn->last_ret = n;
    if (n > 0) return n;
    if (n == 0) return -1;
    switch (SSL_get_error(conn->ssl, n)) {
        case SSL_ERROR_WANT_READ:  return -2;
        case SSL_ERROR_WANT_WRITE: return -3;
        default: return -1;
    }
}

int tls_conn_want_read(tls_conn_t *conn) {
    if (!conn || !conn->ssl) return 0;
    return (SSL_get_error(conn->ssl, conn->last_ret) == SSL_ERROR_WANT_READ) ? 1 : 0;
}

int tls_conn_want_write(tls_conn_t *conn) {
    if (!conn || !conn->ssl) return 0;
    return (SSL_get_error(conn->ssl, conn->last_ret) == SSL_ERROR_WANT_WRITE) ? 1 : 0;
}

/** 取 ALPN 协商结果；buf 为输出缓冲区，返回协议名长度，0 表示未协商或失败。调用方据此判断是否走 h2。 */
int tls_get_alpn_selected(tls_conn_t *conn, char *buf, unsigned int buf_size) {
    if (!conn || !conn->ssl || !buf || buf_size == 0) return 0;
    const unsigned char *out = NULL;
    unsigned int len = 0;
    SSL_get0_alpn_selected(conn->ssl, &out, &len);
    if (!out || len == 0 || len >= buf_size) return 0;
    memcpy(buf, out, len);
    buf[len] = '\0';
    return (int)len;
}

void tls_close(tls_conn_t *conn) {
    if (!conn) return;
    if (conn->ssl) {
        SSL_shutdown(conn->ssl);
        SSL_free(conn->ssl);
        conn->ssl = NULL;
    }
    /* fd 由调用方（Zig）通过 stream.close() 关闭；BIO 模式时 fd 为 -1 */
    conn->fd = -1;
    free(conn);
}

/* ---------- 客户端实现 ---------- */

tls_client_ctx_t *tls_client_ctx_create(const char *ca_path, int verify_peer) {
    const SSL_METHOD *method = TLS_client_method();
    if (!method) return NULL;
    SSL_CTX *ctx = SSL_CTX_new(method);
    if (!ctx) return NULL;
    if (SSL_CTX_set_min_proto_version(ctx, TLS1_2_VERSION) != 1) {
        SSL_CTX_free(ctx);
        return NULL;
    }
    if (verify_peer) {
        SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
        if (ca_path && ca_path[0] != '\0') {
            if (SSL_CTX_load_verify_locations(ctx, ca_path, NULL) != 1 &&
                SSL_CTX_load_verify_locations(ctx, NULL, ca_path) != 1) {
                SSL_CTX_free(ctx);
                return NULL;
            }
        } else {
            if (SSL_CTX_set_default_verify_paths(ctx) != 1) {
                SSL_CTX_free(ctx);
                return NULL;
            }
        }
    } else {
        SSL_CTX_set_verify(ctx, SSL_VERIFY_NONE, NULL);
    }
    tls_client_ctx_t *t = (tls_client_ctx_t *)malloc(sizeof(tls_client_ctx_t));
    if (!t) {
        SSL_CTX_free(ctx);
        return NULL;
    }
    t->ctx = ctx;
    return t;
}

void tls_client_ctx_free(tls_client_ctx_t *t) {
    if (!t) return;
    if (t->ctx) SSL_CTX_free(t->ctx);
    free(t);
}

tls_conn_t *tls_connect(tls_client_ctx_t *t, int fd, const char *servername) {
    if (!t || !t->ctx || fd < 0) return NULL;
    SSL *ssl = SSL_new(t->ctx);
    if (!ssl) return NULL;
    if (SSL_set_fd(ssl, fd) != 1) {
        SSL_free(ssl);
        return NULL;
    }
    if (servername && servername[0] != '\0') {
        if (SSL_set_tlsext_host_name(ssl, servername) != 1) {
            SSL_free(ssl);
            return NULL;
        }
    }
    if (SSL_connect(ssl) <= 0) {
        SSL_free(ssl);
        return NULL;
    }
    tls_conn_t *conn = (tls_conn_t *)malloc(sizeof(tls_conn_t));
    if (!conn) {
        SSL_free(ssl);
        return NULL;
    }
    conn->ssl = ssl;
    conn->fd = fd;
    conn->last_ret = 0;
    return conn;
}

/* ---------- BIO 模式实现 ---------- */

tls_pending_t *tls_accept_start_bio(tls_ctx_t *t) {
    if (!t || !t->ctx) return NULL;
    SSL *ssl = SSL_new(t->ctx);
    if (!ssl) return NULL;
    BIO *rbio = BIO_new(BIO_s_mem());
    BIO *wbio = BIO_new(BIO_s_mem());
    if (!rbio || !wbio) {
        if (rbio) BIO_free(rbio);
        if (wbio) BIO_free(wbio);
        SSL_free(ssl);
        return NULL;
    }
    SSL_set_bio(ssl, rbio, wbio);
    tls_pending_t *p = (tls_pending_t *)malloc(sizeof(tls_pending_t));
    if (!p) {
        SSL_free(ssl);
        return NULL;
    }
    p->ctx = t->ctx;
    p->ssl = ssl;
    p->fd = -1;
    p->last_ret = 0;
    return p;
}

int tls_pending_feed_read(tls_pending_t *pending, const void *buf, int len) {
    if (!pending || !pending->ssl || !buf || len <= 0) return -1;
    BIO *rbio = SSL_get_rbio(pending->ssl);
    if (!rbio) return -1;
    return (BIO_write(rbio, buf, len) > 0) ? 0 : -1;
}

int tls_pending_get_send(tls_pending_t *pending, void *buf, int max_len) {
    if (!pending || !pending->ssl || !buf || max_len <= 0) return 0;
    BIO *wbio = SSL_get_wbio(pending->ssl);
    if (!wbio) return 0;
    int n = BIO_read(wbio, buf, max_len);
    return (n > 0) ? n : 0;
}

int tls_pending_accept_step_bio(tls_pending_t *pending, tls_conn_t **out_conn) {
    if (!pending || !pending->ssl || !out_conn) return -1;
    int r = SSL_accept(pending->ssl);
    pending->last_ret = r;
    if (r == 1) {
        tls_conn_t *conn = (tls_conn_t *)malloc(sizeof(tls_conn_t));
        if (!conn) return -1;
        conn->ssl = pending->ssl;
        conn->fd = -1; /* BIO 模式，无 fd */
        conn->last_ret = 0;
        pending->ssl = NULL;
        free(pending);
        *out_conn = conn;
        return 1;
    }
    switch (SSL_get_error(pending->ssl, r)) {
        case SSL_ERROR_WANT_READ:
        case SSL_ERROR_WANT_WRITE:
            return 0;
        default:
            return -1;
    }
}

int tls_conn_is_bio(tls_conn_t *conn) {
    return (conn && conn->fd == -1) ? 1 : 0;
}

int tls_conn_feed_read(tls_conn_t *conn, const void *buf, int len) {
    if (!conn || !conn->ssl || conn->fd != -1 || !buf || len <= 0) return -1;
    BIO *rbio = SSL_get_rbio(conn->ssl);
    if (!rbio) return -1;
    return (BIO_write(rbio, buf, len) > 0) ? 0 : -1;
}

int tls_conn_read_after_feed(tls_conn_t *conn, void *buf, int len) {
    if (!conn || !conn->ssl || conn->fd != -1 || !buf || len <= 0) return -1;
    int n = SSL_read(conn->ssl, buf, len);
    conn->last_ret = n;
    if (n > 0) return n;
    if (n == 0) return (SSL_get_error(conn->ssl, 0) == SSL_ERROR_ZERO_RETURN) ? 0 : -1;
    switch (SSL_get_error(conn->ssl, n)) {
        case SSL_ERROR_WANT_READ:  return -2;
        case SSL_ERROR_WANT_WRITE: return -3;
        case SSL_ERROR_ZERO_RETURN: return 0;
        default: return -1;
    }
}

int tls_conn_get_send(tls_conn_t *conn, void *buf, int max_len) {
    if (!conn || !conn->ssl || conn->fd != -1 || !buf || max_len <= 0) return 0;
    BIO *wbio = SSL_get_wbio(conn->ssl);
    if (!wbio) return 0;
    int n = BIO_read(wbio, buf, max_len);
    return (n > 0) ? n : 0;
}

int tls_conn_write_app(tls_conn_t *conn, const void *buf, int len) {
    if (!conn || !conn->ssl || conn->fd != -1 || !buf || len <= 0) return -1;
    int n = SSL_write(conn->ssl, buf, (int)len);
    conn->last_ret = n;
    if (n > 0) return n;
    if (n == 0) return -1;
    switch (SSL_get_error(conn->ssl, n)) {
        case SSL_ERROR_WANT_READ:  return -2;
        case SSL_ERROR_WANT_WRITE: return -3;
        default: return -1;
    }
}

File: src/libs/tls/tls.zig
// TLS 服务端封装：options.tls 时在 accept 后做握手，再按明文流读写
// 依赖 C 封装 tls.c（OpenSSL），由 build_options.have_tls 控制是否编译

const std = @import("std");
const build_options = @import("build_options");

/// 当 have_tls 时的 TLS 上下文（cert/key 已加载）
pub const TlsContext = if (build_options.have_tls) struct {
    ptr: *opaque {},
    allocator: std.mem.Allocator,

    /// 从证书与私钥文件路径创建；路径为 UTF-8，失败返回 null
    pub fn create(allocator: std.mem.Allocator, cert_path: []const u8, key_path: []const u8) ?@This() {
        const cert_z = allocator.dupeZ(u8, cert_path) catch return null;
        defer allocator.free(cert_z);
        const key_z = allocator.dupeZ(u8, key_path) catch return null;
        defer allocator.free(key_z);
        const ctx = c.tls_ctx_create(cert_z.ptr, key_z.ptr);
        if (ctx == null) return null;
        return .{ .ptr = @ptrCast(ctx), .allocator = allocator };
    }

    pub fn destroy(self: *@This()) void {
        c.tls_ctx_free(@ptrCast(self.ptr));
        self.ptr = undefined;
    }

    const c = @cImport({
        @cInclude("tls.h");
    });
} else struct {
    pub fn create(_: std.mem.Allocator, _: []const u8, _: []const u8) ?@This() {
        return null;
    }
    pub fn destroy(_: *@This()) void {}
};

/// 当 have_tls 时的客户端 TLS 上下文：CA/校验选项，用于 tls_connect
pub const TlsClientContext = if (build_options.have_tls) struct {
    ptr: *opaque {},
    allocator: std.mem.Allocator,

    /// ca_path 为 CA 证书路径（PEM 文件或目录），null 表示用系统默认；verify_peer 为 true 时验证服务端证书
    pub fn create(allocator: std.mem.Allocator, ca_path: ?[]const u8, verify_peer: bool) ?@This() {
        const ca_z = if (ca_path) |p| allocator.dupeZ(u8, p) catch return null else null;
        defer if (ca_z) |z| allocator.free(z);
        const ca_ptr: [*]const u8 = if (ca_z) |z| z.ptr else @ptrCast("");
        const ctx = c.tls_client_ctx_create(ca_ptr, if (verify_peer) 1 else 0);
        if (ctx == null) return null;
        return .{ .ptr = @ptrCast(ctx), .allocator = allocator };
    }

    pub fn destroy(self: *@This()) void {
        c.tls_client_ctx_free(@ptrCast(self.ptr));
        self.ptr = undefined;
    }

    const c = @cImport({
        @cInclude("tls.h");
    });
} else struct {
    pub fn create(_: std.mem.Allocator, _: ?[]const u8, _: bool) ?@This() {
        return null;
    }
    pub fn destroy(_: *@This()) void {}
};

/// 当 have_tls 时的非阻塞握手中句柄：fd 须已 non-blocking，由 tlsAcceptStart 创建，step 推进或 free 释放
pub const TlsPending = if (build_options.have_tls) struct {
    ptr: *opaque {},
    const c = @cImport({
        @cInclude("tls.h");
    });

    /// 开始非阻塞握手；调用方须已对 fd 调用 setNonBlocking。失败返回 null（调用方需 close(fd)）
    pub fn start(ctx: *const TlsContext, fd: std.posix.socket_t) ?@This() {
        const p = c.tls_accept_start(@ptrCast(ctx.ptr), fd);
        if (p == null) return null;
        return .{ .ptr = @ptrCast(p) };
    }
    /// 查询当前是否需等待可读再 step
    pub fn wantRead(self: *const @This()) bool {
        return c.tls_pending_want_read(@ptrCast(self.ptr)) != 0;
    }
    /// 查询当前是否需等待可写再 step
    pub fn wantWrite(self: *const @This()) bool {
        return c.tls_pending_want_write(@ptrCast(self.ptr)) != 0;
    }
    /// 推进握手：返回 .done 且 out_conn 已写入时调用方取 TlsStream 并 tlsPendingFree；.again 需 poll 后再调；.err 表示失败
    pub fn step(self: *@This(), out_conn: *?*anyopaque) enum { done, again, err } {
        var conn_ptr: ?*anyopaque = null;
        const r = c.tls_accept_step(@ptrCast(self.ptr), @ptrCast(&conn_ptr));
        if (r == 1) {
            out_conn.* = conn_ptr;
            return .done;
        }
        if (r == 0) return .again;
        return .err;
    }
    /// 释放握手中句柄（不关 fd）；握手成功时由 step 内部释放，勿再调
    pub fn free(self: *@This()) void {
        c.tls_pending_free(@ptrCast(self.ptr));
        self.ptr = undefined;
    }
    /// BIO 模式：无 fd，由调用方用 feedRead/getSend 与 overlapped I/O 配合。失败返回 null
    pub fn startBio(ctx: *const TlsContext) ?@This() {
        const p = c.tls_accept_start_bio(@ptrCast(ctx.ptr));
        if (p == null) return null;
        return .{ .ptr = @ptrCast(p) };
    }
    /// BIO 模式：向握手中 SSL 读 BIO 喂入加密数据（WSARecv 完成后调用）
    pub fn feedRead(self: *@This(), buf: []const u8) bool {
        return c.tls_pending_feed_read(@ptrCast(self.ptr), buf.ptr, @intCast(buf.len)) == 0;
    }
    /// BIO 模式：取出待发送加密数据，用于 post WSASend；返回字节数
    pub fn getSend(self: *@This(), buf: []u8) usize {
        const n = c.tls_pending_get_send(@ptrCast(self.ptr), buf.ptr, @intCast(buf.len));
        return if (n > 0) @intCast(n) else 0;
    }
    /// BIO 模式：推进握手；.done 时 out_conn 已写入，pending 已内部释放；.again 需再喂/取后重试
    pub fn stepBio(self: *@This(), out_conn: *?*anyopaque) enum { done, again, err } {
        var conn_ptr: ?*anyopaque = null;
        const r = c.tls_pending_accept_step_bio(@ptrCast(self.ptr), @ptrCast(&conn_ptr));
        if (r == 1) {
            out_conn.* = conn_ptr;
            return .done;
        }
        if (r == 0) return .again;
        return .err;
    }
} else void;

/// 当 have_tls 时的 TLS 流：包装底层 TCP stream，读写经 TLS 加解密
pub const TlsStream = if (build_options.have_tls) struct {
    /// 底层 TCP 连接（close 时由本结构负责关闭）
    underlying: std.Io.net.Stream,
    /// C 层 TLS 连接句柄
    conn: *opaque {},

    /// 从非阻塞握手得到的 C conn 与底层 stream 构造 TlsStream（用于握手完成后；含 BIO 模式 conn）
    pub fn fromConn(underlying: std.Io.net.Stream, conn: *anyopaque) @This() {
        return .{ .underlying = underlying, .conn = @ptrCast(conn) };
    }

    /// 在已 accept 的 stream 上做服务端 TLS 握手（阻塞）；失败返回错误，调用方需 close(stream)
    pub fn accept(underlying: std.Io.net.Stream, ctx: *const TlsContext) !@This() {
        const fd = underlying.socket.handle;
        const conn = c.tls_accept(@ptrCast(ctx.ptr), fd);
        if (conn == null) return error.TlsHandshakeFailed;
        return .{ .underlying = underlying, .conn = @ptrCast(conn) };
    }

    /// 在已连接的 TCP stream 上做客户端 TLS 握手（阻塞）；servername 用于 SNI，可为 null；allocator 用于复制 servername 为 C 字符串
    pub fn connect(underlying: std.Io.net.Stream, client_ctx: *const TlsClientContext, servername: ?[]const u8, allocator: std.mem.Allocator) !@This() {
        const fd = underlying.socket.handle;
        const name_z = if (servername) |s| allocator.dupeZ(u8, s) catch return error.TlsHandshakeFailed else null;
        defer if (name_z) |z| allocator.free(z);
        const name_ptr: [*]const u8 = if (name_z) |z| z.ptr else @ptrCast("");
        const conn = c.tls_connect(@ptrCast(client_ctx.ptr), fd, name_ptr);
        if (conn == null) return error.TlsHandshakeFailed;
        return .{ .underlying = underlying, .conn = @ptrCast(conn) };
    }

    /// ALPN 协商结果：将协议名写入 buf，返回有效切片；未协商或非 h2 时返回 null（供判断是否走 HTTP/2）
    pub fn getAlpnSelected(self: *const @This(), buf: []u8) ?[]const u8 {
        if (buf.len == 0) return null;
        const n = c.tls_get_alpn_selected(@ptrCast(self.conn), buf.ptr, @intCast(buf.len));
        if (n <= 0) return null;
        return buf[0..@intCast(n)];
    }

    /// 实现 Reader：从 TLS 读入明文（阻塞式，内部重试 WANT_*）
    pub fn read(self: *@This(), buf: []u8) !usize {
        const n = c.tls_read(@ptrCast(self.conn), buf.ptr, @intCast(buf.len));
        if (n < 0) return error.TlsReadFailed;
        return @intCast(n);
    }

    /// 非阻塞读：返回字节数或 error.WantRead/error.WantWrite/error.TlsReadFailed；供多路复用 step 使用。BIO 模式时调用方须已 feedRead，此处仅 read_after_feed
    pub fn readNonblock(self: *@This(), buf: []u8) !usize {
        if (self.isBio()) {
            const n = c.tls_conn_read_after_feed(@ptrCast(self.conn), buf.ptr, @intCast(buf.len));
            if (n > 0) return @intCast(n);
            if (n == 0) return 0;
            if (n == -2) return error.WantRead;
            if (n == -3) return error.WantWrite;
            return error.TlsReadFailed;
        }
        const n = c.tls_read(@ptrCast(self.conn), buf.ptr, @intCast(buf.len));
        if (n > 0) return @intCast(n);
        if (n == 0) return 0;
        if (n == -2) return error.WantRead;
        if (n == -3) return error.WantWrite;
        return error.TlsReadFailed;
    }

    /// 非阻塞写：返回已写字节数或 error.WantRead/error.WantWrite/error.TlsWriteFailed；供多路复用 step 使用。BIO 模式时用 writeApp，调用方需随后 getSend 并 post WSASend
    pub fn writeNonblock(self: *@This(), buf: []const u8) !usize {
        if (self.isBio()) {
            const n = c.tls_conn_write_app(@ptrCast(self.conn), buf.ptr, @intCast(buf.len));
            if (n > 0) return @intCast(n);
            if (n == -2) return error.WantRead;
            if (n == -3) return error.WantWrite;
            return error.TlsWriteFailed;
        }
        const n = c.tls_write(@ptrCast(self.conn), buf.ptr, @intCast(buf.len));
        if (n > 0) return @intCast(n);
        if (n == -2) return error.WantRead;
        if (n == -3) return error.WantWrite;
        return error.TlsWriteFailed;
    }

    /// 上次 readNonblock/writeNonblock 返回 WantRead 后，当前是否仍待可读
    pub fn wantRead(self: *const @This()) bool {
        return c.tls_conn_want_read(@ptrCast(self.conn)) != 0;
    }
    /// 上次 readNonblock/writeNonblock 返回 WantWrite 后，当前是否仍待可写
    pub fn wantWrite(self: *const @This()) bool {
        return c.tls_conn_want_write(@ptrCast(self.conn)) != 0;
    }

    /// 是否为 BIO 模式（IOCP 驱动，无 fd 读写）
    pub fn isBio(self: *const @This()) bool {
        return c.tls_conn_is_bio(@ptrCast(self.conn)) != 0;
    }
    /// BIO 模式：向读 BIO 喂入加密数据（WSARecv 完成后调用）；随后可调 readNonblock/readAfterFeed
    pub fn feedRead(self: *@This(), buf: []const u8) bool {
        return c.tls_conn_feed_read(@ptrCast(self.conn), buf.ptr, @intCast(buf.len)) == 0;
    }
    /// BIO 模式：喂入后读明文；语义同 readNonblock（>0 字节数，0 关闭，WantRead/WantWrite/TlsReadFailed）
    pub fn readAfterFeed(self: *@This(), buf: []u8) !usize {
        const n = c.tls_conn_read_after_feed(@ptrCast(self.conn), buf.ptr, @intCast(buf.len));
        if (n > 0) return @intCast(n);
        if (n == 0) return 0;
        if (n == -2) return error.WantRead;
        if (n == -3) return error.WantWrite;
        return error.TlsReadFailed;
    }
    /// BIO 模式：取出待发送加密数据；返回字节数，用于 post WSASend
    pub fn getSend(self: *@This(), buf: []u8) usize {
        const n = c.tls_conn_get_send(@ptrCast(self.conn), buf.ptr, @intCast(buf.len));
        return if (n > 0) @intCast(n) else 0;
    }
    /// BIO 模式：应用层写明文；语义同 writeNonblock。可能使写 BIO 有数据，需 getSend 取出后 WSASend
    pub fn writeApp(self: *@This(), buf: []const u8) !usize {
        const n = c.tls_conn_write_app(@ptrCast(self.conn), buf.ptr, @intCast(buf.len));
        if (n > 0) return @intCast(n);
        if (n == -2) return error.WantRead;
        if (n == -3) return error.WantWrite;
        return error.TlsWriteFailed;
    }

    /// 实现 Writer：向 TLS 写入明文（阻塞式，内部重试 WANT_*）
    pub fn writeAll(self: *@This(), buf: []const u8) !void {
        var off: usize = 0;
        while (off < buf.len) {
            const n = c.tls_write(@ptrCast(self.conn), buf.ptr + off, @intCast(buf.len - off));
            if (n <= 0) return error.TlsWriteFailed;
            off += @intCast(n);
        }
    }

    /// 关闭 TLS 并关闭底层 stream。0.16：underlying.close(io)
    pub fn close(self: *@This(), io: std.Io) void {
        c.tls_close(@ptrCast(self.conn));
        self.conn = undefined;
        self.underlying.close(io);
    }

    const c = @cImport({
        @cInclude("tls.h");
    });
} else struct {
    underlying: std.Io.net.Stream,
    pub fn accept(_: std.Io.net.Stream, _: *const TlsContext) !@This() {
        return error.TlsNotCompiled;
    }
    pub fn read(_: *@This(), _: []u8) !usize {
        return error.TlsNotCompiled;
    }
    pub fn writeAll(_: *@This(), _: []const u8) !void {
        return error.TlsNotCompiled;
    }
    pub fn close(_: *@This(), _: std.Io) void {}
};
